CRM-19678: Create membership signup activity instead of Membership Renewal on Pay later

monishdeb opened the pull request:
* [CRM-19678: No Membership Renewal Activity is created when a Pay Later is set to Completed](https://issues.civicrm.org/jira/browse/CRM-19678)

monishdeb commented:
@petednz I have fixed the issue where earlier it create Membership Renewal Activity on doing Pay later over current membership. Added unit test to assert that Membership Renewal is only created on Completed contribution.

eileenmcnaughton commented:
Hmm - what about delayed payment processor scenario - which is not pay later but does not happen in real time. A payment processor might take days or weeks to update. We have been deliberately eroding the concept of paylater in order to recognise the fact that a payment processor may well behave in the same way as a 'pay-later'

 Can we not create the activity in the BAO based on whether it is a new membership or an update to an existing one (extending dates).

monishdeb commented:
@eileenmcnaughton we are already creating activities inside ```CRM_Member_BAO_Membership::create(...)``` based on new or updating i.e membership ID [here](https://github.com/monishdeb/civicrm-core/blob/47accb98315e6c9f797246457a7847f83ac1b89f/CRM/Member/BAO/Membership.php#L416) But on further digging, it seems like there is a possibility to bring the change (decide Activity type) in ```CRM_Member_BAO_Membership::create(...)``` as membership ```$params``` carry 'contribution' attribute and we can use this ```$params['contribution']->status_id``` information to create Activity of desired type. But then we cannot ensure that the membership params will contain the contribution information always to decide the fate of related activity. One of the reason why they have predetermined activity type as the 3rd parameter - [$actvityType](https://github.com/monishdeb/civicrm-core/blob/47accb98315e6c9f797246457a7847f83ac1b89f/CRM/Member/BAO/Membership.php#L232) of ```Membership::create(...) instead of doing inside it BAO function.
 
Now about the fix, earlier for current membership renewal we simply ignored related activity's type [here](https://github.com/monishdeb/civicrm-core/blob/47accb98315e6c9f797246457a7847f83ac1b89f/CRM/Member/BAO/Membership.php#L1857) whether the payment has been made or not. So I have changed the type to 'Membership Signup' IF the contribution is pending for pay-later ONLY. If I remove the IF condition then it will simply mean we creating 'Membership Signup' always in case of pending contribution due to pay-later or (like you said) incomplete payment or failed payment.

eileenmcnaughton commented:
So in here

      if (!empty($ids['membership']) && $activityType != 'Membership Signup') {
        CRM_Activity_BAO_Activity::addActivity($membership, $activityType, $targetContactID);
      }
      elseif (empty($ids['membership'])) {
        CRM_Activity_BAO_Activity::addActivity($membership, $activityType, $targetContactID);
      }

Why wouldn't we go in here https://github.com/monishdeb/civicrm-core/blob/47accb98315e6c9f797246457a7847f83ac1b89f/CRM/Member/BAO/Membership.php#L142 and create the membership signup activity from there if creating a new membership?

monishdeb commented:
@eileenmcnaughton done and here's my changes https://github.com/civicrm/civicrm-core/pull/9468/commits/9ffe1b03d2b6457eab2bb983d738f3bcf230f9ef 

1. Removed explicit ```$activityType``` third parameter for ```Membership::create(...)``` [here](https://github.com/civicrm/civicrm-core/pull/9468/commits/ae95735cd85e9537138fc4d3700f8a3f8668cd98#diff-f43c8498e32f5b2d68ab27bcd243ca36L238)
2. Moved ```addActivity()``` call into ```Membership::add(...)``` based on ```id``` [here](https://github.com/civicrm/civicrm-core/pull/9468/commits/ae95735cd85e9537138fc4d3700f8a3f8668cd98#diff-f43c8498e32f5b2d68ab27bcd243ca36R175)
3. Handled all 4 scenarios in unit test [here](https://github.com/civicrm/civicrm-core/pull/9468/commits/ae95735cd85e9537138fc4d3700f8a3f8668cd98#diff-40e2e0f106ba620465acf3a9a81f2498R2566)

monishdeb commented:
Although found another issue where on membership renewal via online contribution page, 3 activities are created 
![image](https://cloud.githubusercontent.com/assets/3735621/20797986/35dcb08a-b803-11e6-8d90-ff33952ca647.png)

expected behaviour - Renewal and Contribution Activity

monishdeb commented:
Jenkin test this please

monishdeb commented:
Jenkins, test this please

eileenmcnaughton commented:
I did a quick skim and it looks like you tackled some pretty scary code & made it cleaner. I think it will need a bit more time to read the code properly though

Good score given that you added tests

petednz commented:
To avoid confusion should this ticket be retitled to be clear this is for when a New Membership is being created?

monishdeb commented:
@petednz I agree it just add confusion but the overall improvement is for greater good :) The changes ensure that there always be a renewal activity created on membership.update BAO fn and membership signup activity on  membership.create by shifting the ```addActivity``` call to BAO fn rather then calling explicitly at other places as per need. Thanks @eileenmcnaughton for the suggestion. Also the patch include other optimization

petednz commented:
no problem, i thought this was still referring to when a Pay Later is set to Completed, but take it this is for the previous step. thanks

monishdeb commented:
Yes it covers payLater to Completed use-case too as I confirmed in my local to check if right activity(s) are created after each process and then there is a unit test to ensure my result and other 3 use-cases.

eileenmcnaughton commented:
@petednz if you can test & confirm the patch I can merge. I think @Stoob did some instructions on how to get to rc demo sites?

petednz commented:
So let me try and clarify what are the expected outcomes since it seems to have shifted
- when an online Membership form is submitted using Pay Later, 
-- a Signup activity is created? and set to Completed? or Pending?
-- a Renewal Activity is created? and set to Pending?
- when the Contribution is set to Completed
-- the Signup Activity is set to Completed?
-- the Renewal Activity is set to Completed? or is the Renewal Activity created at this point?

monishdeb commented:
I am afraid as per earlier workflow we are not changing the Activity status but always create a Completed one of desired Activity type. I have just shifted the call to BAO fn nothing to manipulate the Activity status https://github.com/civicrm/civicrm-core/pull/9468/files#diff-a35044abd0f538e4af551c5ae039692dR1829

Stoob commented:
I'm coming at this late I realize.  But there is no 'pending' activity 
status.  And given that the default Activity search is both "scheduled" 
and "completed" I think adding activities before the contribution is 
complete will confuse reporting.

I don't think a signup/renewal status needs to exist at all until the 
contribution is completed.



On 12/04/2016 7:56 pm, Monish Deb wrote:
> I am afraid as per earlier workflow we are not changing the Activity
> status but always create a Completed one of desired Activity type. I
> have just shifted the call to BAO fn nothing to manipulate the
> Activity status
> https://github.com/civicrm/civicrm-core/pull/9468/files#diff-a35044abd0f538e4af551c5ae039692dR1829
> 
> 
> --
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub [1], or mute the
> thread [2].
> 
> 	*
> 
> Links:
> ------
> [1] 
> https://github.com/civicrm/civicrm-core/pull/9468#issuecomment-264762931
> [2]
> https://github.com/notifications/unsubscribe-auth/ABmwEQgPphETSQYqhhGzx82z50a7JTZ6ks5rE4uEgaJpZM4K-68M

monishdeb commented:
sorry :( ya I meant 'scheduled' (overlooked the available activity statuses)

Stoob commented:
Thanks Monish.  Can you confirm with Pete that an activity should exist 
prior to the contribution completed?  I'd offer that perhaps the trigger 
for creating an 'membership signup/renewal' activity should be a 
completed contribution.

My Civiteacher.com site dating back to early 4.4 has no incidents of 
'scheduled' 'membership signup/renewal' activities, so there is no 
precdent for these activities.  These activities have historically 
always been status:completed, and always created when the contribution 
is complete.


On 12/04/2016 8:07 pm, Monish Deb wrote:
> sorry :( ya I meant 'scheduled' (overlooked the available activity
> statuses)
> 
> --
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub [1], or mute the
> thread [2].
> 
> 	*
> 
> Links:
> ------
> [1] 
> https://github.com/civicrm/civicrm-core/pull/9468#issuecomment-264763825
> [2]
> https://github.com/notifications/unsubscribe-auth/ABmwEUjzOL3dy9adxf4dNG-ce4j7xKdEks5rE43vgaJpZM4K-68M

monishdeb commented:
What if
1.  on pending(pay-later) to completed online registration membership : Scheduled Membership Signup (with required ```details```) to Completed Membership Signup  
2. on completed to renewal : Completed Membership Signup to Renewal Membership activity 
3. completed to pay-later to renewal : Completed Membership Signup to Scheduled Membership Renewal to Completed Membership Renewal

So basically as per each action : 
-  ```pay-later or pending membership``` it will always create schedule membership signup and/or membership renewal irrespective of backoffice or online registration
-  ```completed membership``` - create/update Complete membership signup activity
-  ```renew membership``` - create/update Complete membership renew activity 

(And this proposed changes will be from 4.7 until unless there is need to backport)

petednz commented:
I think it is useful to know when a Renewal was submitted with a Pending Pay Later.

If there is nothing that does that currently (such as an Activity), then adding a 'Scheduled' Activity seems like useful information (ie I can go search for 
- Activity Type = Membership Renewal + Activity Status = Scheduled

Stoob commented:
The only way currently to search now would be a pending pay-later 
"member dues" contribution.


On 12/08/2016 2:27 pm, petednz wrote:
> I think it is useful to know when a Renewal was submitted with a
> Pending Pay Later.
> 
> If there is nothing that does that currently (such as an Activity),
> then adding a 'Scheduled' Activity seems like useful information (ie I
> can go search for
> 
>  	* Activity Type = Membership Renewal + Activity Status = Scheduled
> 
> --
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub [1], or mute the
> thread [2].
> 
> 	*
> 
> Links:
> ------
> [1] 
> https://github.com/civicrm/civicrm-core/pull/9468#issuecomment-265874816
> [2]
> https://github.com/notifications/unsubscribe-auth/ABmwEfh3DwcLSdzw8ExcVVv22lz_njiCks5rGIRPgaJpZM4K-68M

petednz commented:
which by itself wouldn't differentiate between New v Renew?

Stoob commented:
Nope, you're right it wouldn't


On 12/08/2016 2:48 pm, petednz wrote:
> which by itself wouldn't differentiate between New v Renew?
> 
> --
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub [1], or mute the
> thread [2].
> 
> 	*
> 
> Links:
> ------
> [1] 
> https://github.com/civicrm/civicrm-core/pull/9468#issuecomment-265879282
> [2]
> https://github.com/notifications/unsubscribe-auth/ABmwEXnzbzF0NWy5eI5fwYC-dE7fdelnks5rGIlbgaJpZM4K-68M

petednz commented:
Monish - I am still struggling to understand your breakdown so excuse me while i try and rephrase for my own benefit. First, i don't understand why you say 

3: completed to pay-later to renewal : Completed Membership Signup to Scheduled Membership Renewal to Completed Membership Renewal

I do not think there should be a 'Membership Signup' when a renewal is happening. I think it should only happen when a Membership is being created.

So my sense the options would be
New membership + Pay Now > Membership Sign up (Completed)
New membership + Pay Later  > Membership Sign up (Scheduled)
Renew + Pay Now  > Membership Renewal (Completed)
Renew + Pay Later  > Membership Renewal (Scheduled)

monishdeb commented:
@petednz sorry for using misnomer ```Completed``` instead of ```New Membership```. Lemme rephrase my 3rd point. 
>  New Membership to pay-later to Membership Renewal : Completed Membership Signup to Scheduled Membership Renewal to Completed Membership Renewal activities respectively 

What I meant was 3 sequential steps:
1. At first when user does a new membership then  Completed Membership Signup activity  
2. Next when he wants to renew his membership but chose Pay-later then Scheduled Membership Renewal
3. On successful renewal - Completed Membership Renewal (basically status change of previous scheduled membership) 

And yes those are the options what I meant in my previous comment. So I'll update my PR with these changes and with proper test coverage of all possible use-cases :)

petednz commented:
i think with your 3 sequential steps i am not clear if you need to also refer to the New Member this is Pay Later, or not, because I have really lost track of what you are endeavouring to add/fix/change ;-)

monishdeb commented:
Ummm...  
New Member = (New membership + Pay Now) =  Membership Sign up (Completed)

monishdeb commented:
Jenkin test this please

monishdeb commented:
Jenkin ok to test

eileenmcnaughton commented on CRM/Member/BAO/Membership.php:
```diff
@@ -139,38 +139,65 @@ public static function add(&$params, $ids = array()) {
 
+    $activityParams = array();
+    $allStatus = CRM_Member_BAO_Membership::buildOptions('status_id', 'get');
+    if (!empty($params['is_pay_later']) ||
```
Working through this function I think the question is when would we want to create the activity and for it NOT to be pending/ scheduled. It seems like the old behaviour was for it to always be created as Scheduled, when created. So, why don't we add a parameter 'membership_activity_status_id' that can be set to Completed if desired, but we default to scheduled? That way we aren't trying to deal with the BAO function deciding how pay later it is. And, the core form functions should always create pending and then call complete transaction, unless there is no payment at all I think. Getting away from is_pay_later & making the calling function specify the activity_status id seems more explicit & less random.

My other thinking on this was that the place where we REALLY know what is going on is in the MembershipPayment::create function. But I can't quite see how to turn that nugget into a strategy

eileenmcnaughton replied:
I guess one thing we should note is that in 4.7 we made it so that the forms should always create Pending contributions & then use the completetransaction api to complete them when the payment is complete (by whatever method). The activity creation code was never updated to reflect that, which I think it making it confusing to think through.

I think we can probably create scheduled activities from that form on the basis that the form should follow the 4.7 flow?

So, what about the membership BAO accepts a parameter 'membership_activity_status_id' => 'Scheduled' & we set it to that when coming in off that form flow & it can default to Completed in other scenarios (at least for now) ?

eileenmcnaughton commented on CRM/Member/BAO/Membership.php:
```diff
@@ -1876,11 +1853,13 @@ public static function processMembership($contactID, $membershipTypeID, $is_test
           'join_date' => $currentMembership['join_date'],
           'membership_type_id' => $membershipTypeID,
           'max_related' => !empty($membershipTypeDetails['max_related']) ? $membershipTypeDetails['max_related'] : NULL,
+          'is_pay_later' => $isPayLater,
```
per comments above - let's not propagate this parameter - if we need to control the membership_activity_status_id let's call it that.

eileenmcnaughton commented on tests/phpunit/api/v3/ContributionTest.php:
```diff
@@ -2485,12 +2485,6 @@ public function testCompleteTransactionMembershipPriceSet() {
     $logs = $this->callAPISuccess('MembershipLog', 'get', array(
       'membership_id' => $this->_ids['membership'],
     ));
-    //CRM-19600: Ensure that 'Membership Renewal' activity is created after successful membership regsitration
```
Can you clarify this for me. I'm guessing you have decided the test was testing behaviour that was incorrect?

monishdeb replied:
Earlier after the successfully payment of a pending membership, membership renewal was created which was incorrect. As per the fix https://github.com/civicrm/civicrm-core/pull/9468/files#diff-7a5b0e2d131dabc49178460fc63328c0L4603 the expected behaviour would be to change the 'Scheduled' Membership Signup Activity to 'Completed'. Reason why I have removed this test-case.

eileenmcnaughton replied:
OK cool

eileenmcnaughton commented on tests/phpunit/api/v3/ContributionTest.php:
```diff
@@ -2507,6 +2501,13 @@ public function testPendingToCompleteContribution() {
     );
     $this->setUpPendingContribution($this->_ids['price_field_value'][0]);
     $this->callAPISuccess('membership', 'getsingle', array('id' => $this->_ids['membership']));
+    // Case 1: Assert that Membership Signup Activity is created on Pending (Pay later) to Completed Contribution via backoffice
```
Remove words pay later - the contribution created is (specifically) not a pay later one

eileenmcnaughton commented on tests/phpunit/api/v3/ContributionTest.php:
```diff
@@ -2548,11 +2549,29 @@ public function testPendingToCompleteContribution() {
     catch (Civi\Payment\Exception\PaymentProcessorException $e) {
       $error = TRUE;
     }
+    // Case 2: After successful payment for Pending (Pay later) backoffice there are three activities created
```
Remove word pay Later

eileenmcnaughton commented:
test this please

eileenmcnaughton commented:
I've mad some comments from a UI point of view but @petednz or @Stoob would need to test this from a UI point of view before we merge

monishdeb commented:
Strange that it never show the test build. The is the current test build https://test.civicrm.org/job/CiviCRM-Core-PR/13076/

monishdeb commented on CRM/Contribute/BAO/Contribution.php:
```diff
@@ -562,25 +562,10 @@ public static function create(&$params, $ids = array()) {
       $contribution->find(TRUE);
     }
-    CRM_Activity_BAO_Activity::addActivity($contribution, 'Offline');
+    CRM_Activity_BAO_Activity::addActivity($contribution, 'Contribution');
```
In fact there is no ```Offline``` activity type ..lol

monishdeb commented:
@eileenmcnaughton I have made all the changes as you suggested. Please have a look :)

petednz commented:
anything we can do to help here?

eileenmcnaughton commented:
@petednz sure I spoke with Gemma - but this needs someone like you or Gemma to review. Code seems fine but needs to be tested against requirements

eileenmcnaughton commented:
@petednz This still seems stalled on user testing - were you going to sort that?

petednz commented:
@jitendrapurohit can i get you to apply these changes to fuzzy site so we can pls test. ping me when suits

petednz commented:
Kaylie did a test with a contact that was created in back end, then renewed via a contribution page using pay later option.
After renewal: Activity "Membership Renewal" was created with Subject "General - Status: Grace" and Status "Schedule"
![image](https://cloud.githubusercontent.com/assets/5800574/24438375/7a8a2cce-14a3-11e7-8743-ee4abd9cdd7d.png)
Contribution was then changed from "Pending - Pay Later" to "Completed" through Contributions.
A new activity was then created. Activity "Membership Renewal", Subject "General - Status" Current", Status "Completed".
This left the contact with two Membership Renewal activities for a single membership renewal.
![image](https://cloud.githubusercontent.com/assets/5800574/24438382/8630ad28-14a3-11e7-93c0-96ff7de336a1.png)

The outcome should be that the Renewal activity should be updated from 'Scheduled' to 'Completed', not a second activity created

petednz commented:
Kaylie did secondary test. Contact membership created in back end, then renewed via the back end, selecting "Paid by Check", Contribution Status "Completed".
After renewal: Activity "Membership Renewal" created. Subject "General - Status" Current. Status - no status.
![image](https://cloud.githubusercontent.com/assets/5800574/24438690/86b51b7e-14a5-11e7-8e14-46af82daa3c6.png)

The outcome should be that the Renewal activity should have status of 'Completed'

petednz commented:
There is summary sheet here of the outcomes we are finding after applying this patch

https://docs.google.com/spreadsheets/d/1UPfA10UIKjuOnTqyradi6vfOhveUts161PpSm_sb5f0/edit#gid=0

monishdeb commented:
Closed in favor of https://github.com/civicrm/civicrm-core/pull/10323
